src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -245,12 +246,102 @@ private BoundExpression ApplyConvertedTypes(BoundExpression expr, TupleBinaryOpe
             bool rightNullable = right.Type?.IsNullableType() == true;
             bool isNullable = leftNullable || rightNullable;
 
-            TypeSymbol leftTupleType = MakeConvertedType(operators.SelectAsArray(o => o.LeftConvertedTypeOpt), node.Left, leftParts, isNullable, compilation, diagnostics);
-            TypeSymbol rightTupleType = MakeConvertedType(operators.SelectAsArray(o => o.RightConvertedTypeOpt), node.Right, rightParts, isNullable, compilation, diagnostics);
+            TypeSymbol leftTupleType = MakeConvertedType(operators.SelectAsArray(o => o.LeftConvertedTypeOpt), node.Left, leftParts, leftNames, isNullable, compilation, diagnostics);
+            TypeSymbol rightTupleType = MakeConvertedType(operators.SelectAsArray(o => o.RightConvertedTypeOpt), node.Right, rightParts, rightNames, isNullable, compilation, diagnostics);
 
             return new TupleBinaryOperatorInfo.Multiple(operators, leftTupleType, rightTupleType);
         }
 
+        /// <summary>
+        /// If an element in a tuple literal has an explicit name which doesn't match the name on the other side, we'll warn.
+        /// The user can either remove the name, or fix it.
+        ///
+        /// This method handles two expressions, each of which is either a tuple literal or an expression with tuple type.
+        /// In a tuple literal, each element can have an explicit name, an inferred name or no name.
+        /// In an expression of tuple type, each element can have a name or not.
+        /// </summary>
+        private static void ReportNamesMismatchesIfAny(BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
+        {
+            bool leftIsTupleLiteral = left.Kind == BoundKind.TupleLiteral;
+            bool rightIsTupleLiteral = right.Kind == BoundKind.TupleLiteral;
+
+            if (!leftIsTupleLiteral && !rightIsTupleLiteral)
+            {
+                return;
+            }
+
+            var (leftNames, leftInferred, leftLocations) = getDetails(left);
+            bool leftNoInferredNames = leftInferred.IsDefault;
+            bool leftNoNames = leftNames.IsDefault;
+
+            var (rightNames, rightInferred, rightLocations) = getDetails(right);
+            bool rightNoInferredNames = rightInferred.IsDefault;
+            bool rightNoNames = rightNames.IsDefault;
+
+            if (leftNoNames && rightNoNames)
+            {
+                return;
+            }
+
+            Debug.Assert(leftNoNames || rightNoNames || leftNames.Length == rightNames.Length);
+
+            // The only case where the right side is not preferred is when we have a tuple literal on the left, but not the right
+            bool isRightBestSide = !(leftIsTupleLiteral && !rightIsTupleLiteral);
```

Comment:
Can this be simplified to `bool isRightBestSide = rightIsTupleLiteral;` since we know at least one of the two are literals.